src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs

```diff
@@ -4,46 +4,80 @@
-                                                                             s_localizableTitle,
-                                                                             s_localizableMessageNotIntegral,
-                                                                             DiagnosticCategory.Design,
-                                                                             DiagnosticSeverity.Warning,
-                                                                             isEnabledByDefault: false,
-                                                                             description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182147.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(NotInt32Rule, NotIntegralRule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            {
+                var flagsAttribute = WellKnownTypes.FlagsAttribute(compilationContext.Compilation);
+                compilationContext.RegisterSymbolAction(symbolContext => AnalyzeSymbol(symbolContext, flagsAttribute), SymbolKind.NamedType);
+            });
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol flagsAttribute)
+        {
+            var symbol = (INamedTypeSymbol)context.Symbol;
+
+            if (symbol.TypeKind != TypeKind.Enum)
+            {
+                return;
+            }
+
+            var underlyingType = symbol.EnumUnderlyingType.SpecialType;
+            if (underlyingType == SpecialType.System_Int32)
+            {
+                return;
+            }
+
+            // If accessibility of enum is not public exit
+            if (symbol.GetResultantVisibility() != SymbolVisibility.Public)
+            {
+                return;
+            }
+
+            // If enum is Int64 and has Flags attributes then exit
+            bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
+            if (underlyingType == SpecialType.System_Int64 && hasFlagsAttribute)
+            {
+                return;
+            }
             
+            //Report diagnostic
+            var diagnostic = Diagnostic.Create(Rule, symbol.Locations[0], symbol.Name, symbol.EnumUnderlyingType.ToString());
```

Comment:
Here's where you can just say:
```
symbol.CreateDiagnostic(Rule, symbol.Name, symbol.EnumUnderlyingType);
```
Take a look at src\Analyzer.Utilities\Extensions.cs. You'll see lots of convenient extension methods named `CreateDiagnostic` or `CreateDiagnostics`. You can call `CreateDiagnostic` on a `SyntaxNode` as well as a symbol, because again, the node knows its own location.